File: medium/cpp/974.cpp
// https://leetcode.com/problems/subarray-sums-divisible-by-k/

class Solution
{
public:
    int subarraysDivByK(std::vector<int>& nums, int k)
    {
        std::vector<int> counts(k);
        counts[0] = 1;
        
        int sum = 0;
        int ans = 0;
        
        for (auto e : nums)
        {
            sum += e;
            sum = (sum % k + k) % k;
            
            ans += counts[sum];
            
            counts[sum]++;
        }
        
        return ans;
    }
};

File: medium/python/974.py
# https://leetcode.com/problems/subarray-sums-divisible-by-k/

class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        counts = [0] * k
        counts[0] = 1
        
        ans = 0
        sum = 0
        
        for num in nums:
            sum += num
            sum = (sum % k + k) % k
            
            ans += counts[sum]
            counts[sum] += 1
            
        return ans
        